### DISCUSSION TITLE
NextJS 13, Pocketbase Auth and middleware.

### DISCUSSION THREAD

**Discussion by rlawrence89** [pocketbase/pocketbase#2939] [Q&A] [answered]
Can someone explain how i can auth guard my app using Next JS 13 middleware? My app would be using a mix of SSR and CSR. I have not come across any kind of solution so far.

**Comment by tigawanna** [comment]
Here's how I do it with the app router [article](https://dev.to/tigawanna/nextjs13-app-directory-authentication-with-pocketbase-2p6d]

**Reply by rlawrence89** [reply]
Thanks for this. Will definitely test it out. But from what i can see the alow of the middleware is such:

1. Extract the "pb_auth" from the request cookie.
2. Validate the cookie and refresh it.
3. Check the authstore for a user model. 

Simplification from what i can gather from the middleware code. I also looked at your appwrapper. Can you explain the use of canUseCookie?

**Reply by tigawanna** [reply]
That's the flow , the complication with Nextjs is that they don't have a good way to pass an object between requests, and you can't write to a cookie in the server component unless if you use server actions, 
The can use cookie was a helper that checked if the user declined cookies , it's unnecessary if you're not using 3d party cookies

**Reply by rlawrence89** [reply]
Thanks alot, I have made significant progress. I am able to log in and navigate. Now i am attempting to logout. It clears the local storage but if i hit refresh i am still logged in and authstore is placed back in localstorage but with null values.

Here is my middleware file. 
```
import { NextResponse } from 'next/server'
import { cookies } from 'next/headers'
import type { NextRequest } from 'next/server'
import pb from './utils/pocketbase'

// This function can be marked `async` if using `await` inside
export async function middleware(request: NextRequest) {
  // console.log('Document Cookie:', document?.cookie)
  const response = NextResponse.next()
  const request_cookie = request.cookies.get('pb_auth')

  if (request_cookie || '') {
    try {
      console.log('Loadinng cookie:', request_cookie)

      const cookie = { pb_auth: request_cookie?.value || '' }
      let encodedCookie = ''
      for (const [key, value] of Object.entries(cookie)) {
        encodedCookie += `${encodeURIComponent(key)}=${encodeURIComponent(
          value,
        )}; `
      }
      console.log('Encoded Cookie: ', encodedCookie)
      pb.authStore.loadFromCookie(encodedCookie.trimEnd())
      // console.log('Model: ', pb.authStore)
    } catch (error) {
      console.log('Error Loading from cookie: ', error)
      pb.authStore.clear()
      response.headers.set(
        'set-cookie',
        pb.authStore.exportToCookie({ httpOnly: false }),
      )
    }
  }

  try {
    // get an up-to-date auth store state by verifying and refreshing the loaded auth model (if any)
    pb.authStore.isValid && (await pb.collection('users').authRefresh())
    console.log('authstore is valid')
    // console.log('authstore:', pb.authStore)
  } catch (err) {
    console.log('Invalid authstore: ', err)
    // clear the auth store on failed refresh
    pb.authStore.clear()
    response.headers.set(
      'set-cookie',
      pb.authStore.exportToCookie({ httpOnly: false }),
    )
  }

  // console.log('Model: ', pb.authStore)
  if (!pb.authStore.model && !request.nextUrl.pathname.startsWith('/login')) {
    const redirect_to = new URL('/login', request.url)
    if (request.nextUrl.pathname) {
      redirect_to.search = new URLSearchParams({
        next: request.nextUrl.pathname,
      }).toString()
    } else {
      redirect_to.search = new URLSearchParams({
        next: '/',
      }).toString()
    }
    console.log('No Model and not on login. Redirectinng to: ', redirect_to)
    return NextResponse.redirect(redirect_to)
  }

  if (pb.authStore.model && request.nextUrl.pathname.startsWith('/login')) {
    const next_url = request.headers.get('next-url') as string
    if (next_url) {
      const redirect_to = new URL(next_url, request.url)
      return NextResponse.redirect(redirect_to)
    }
    const redirect_to = new URL(`/`, request.url)
    return NextResponse.redirect(redirect_to)
  }

  return response
}

export const config = {
  //Regex used to allow nextjs routes through
  matcher: ['/((?!api|_next/static|_next/image|favicon.ico).*)'],
}

``` 

Here is my Logout button

```
'use client'
import pb from '@/utils/pocketbase'
import { useRouter } from 'next/navigation'
import React from 'react'

type Props = {}

const Logout = (props: Props) => {
  const router = useRouter()
  function logoutHandler() {
    pb.authStore.clear()
    document.cookie = pb.authStore.exportToCookie({ httpOnly: false })
    router.refresh()
  }
  return (
    <button
      onClick={() => logoutHandler()}
      className="rounded-md bg-red-600 p-2 text-white"
    >
      Logout
    </button>
  )
}

export default Logout

```

And ive adapted your appWraper 
```
'use client'

import pb from '@/utils/pocketbase'
import React, { useEffect } from 'react'

type Props = { children: React.ReactNode }

const AuthListener = (props: Props) => {
  useEffect(() => {
    // console.log("user changed  ===", pb.authStore.model);
    pb.authStore.loadFromCookie(document?.cookie ?? '')
    const authChange = pb.authStore.onChange((token, model) => {
      console.log('AuthListener:', token, model)
      document.cookie = pb.authStore.exportToCookie({ httpOnly: false })
    })
    return () => {
      //  call to clean up the listener
      authChange()
    }
  }, [])
  return <>{props.children}</>
}

export default AuthListener

```

Not sure why its not logging out properly.

**Comment by tigawanna** [comment, marked as answer]
your logout logic looks very similar to mine the general idea is calling 
```ts
        pb.authStore.clear();
        document.cookie=pb.authStore.exportToCookie({ httpOnly: false });
then router.refresh()
```
which clears out the local storage `authstore` and updates the cookie to an empty string

looking at your middleware you're importing a PocketBase instance from somewhere else am guessing it's from the instance you're using client side
 ```ts
import pb from "./utils/pocketbase";
```
The recommended way is to create a new instance on every request and use that instance to avoid unpredictable behavior
```ts
  const pb = new PocketBase(pb_url);
```

then after you hit the logout button , check both application tab in `devtools` for the `pb_auth`` cookie and local storage values ,they should be empty , if NextJS isn't redirecting you to the login page the issue could in the middleware matcher

```ts
export const config = {
  //Regex used to allow nextjs routes through
  matcher: ['/((?!api|_next/static|_next/image|favicon.ico).*)'],
}

```

**Reply by rlawrence89** [reply]
Thanks for all your help. I believe the creation of a new PB instance is what allowed it to work.

**Reply by plantshopping** [reply]
> The recommended way is to create a new instance on every request
Does this mean that everytime we need to use pocketbase we need to create a new instance? And we should not use a client?

My pb client for example:
```
import PocketBase from "pocketbase";

export const POCKET_BASE_URL = "http://127.0.0.1:8090";

export class DatabaseClient {
  client: PocketBase;

  constructor() {
    this.client = new PocketBase(POCKET_BASE_URL);
  }

 // other methods here  
}

export const db = new DatabaseClient();

export default db;
```

**Reply by ganigeorgiev** [reply]
@plantshopping The above seems vulnerable for the reason explained in "Problem 1" in https://github.com/pocketbase/pocketbase/discussions/5313 (imported modules are executed only once; additionally having both named and default export pointing to the same thing doesn't really make much sense unless supporting both cases was the intended behavior). In any case, if you are new to web development I strongly advice to read the linked discussion and avoid using JS SSR with PocketBase if you are not sure that you need it and where and how your code is running.

**Comment by tigawanna** [comment]
Glad to help
